streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala

```diff
@@ -974,6 +977,7 @@ class SlowTestReceiver(totalRecords: Int, recordsPerSecond: Int)
     }
     receivingThreadOption = Some(thread)
     thread.start()
+    SlowTestReceiver.initialized = true
```

Comment:
Shouldn't this be in the `Thread.run()` implementation above?

Reply:
If we consider only timing, technically it can be placed anywhere, even first line of `onStart()`, as whether receiver is registered within time or not is the key. For verification, it doesn't make existing test failing without patch even we add `Thread.sleep(1000)` in first line of `onStart()`.

So no strong opinion on where to put. 

Btw, maybe we can apply "more verbose but clearer" solution (without this flag) via adding below code in test side:

```
    // tracks whether the receiver is started or not
    var isReceiverStarted = false
    val listener = new StreamingListener {
      override def onReceiverStarted(receiverStarted: StreamingListenerReceiverStarted): Unit = {
        isReceiverStarted = receiverStarted.receiverInfo.name.startsWith(
          input.getReceiver().getClass.getSimpleName)
      }
    }
    ssc.addStreamingListener(listener)

    ssc.start()
    ssc.awaitTerminationOrTimeout(500)

    eventually(timeout(10.seconds), interval(10.millis)) {
      assert(isReceiverStarted)
    }
```

Which one do you think is better?

Reply:
> as whether receiver is registered within time or not is the key

Ah ok. Was wondering if the thread that actually does stuff needed to run for this to work, but if it's just the registration that matters, this is enough.

But can't the flag be in the actual `SlowTestReceiver` instance (instead of the new object)?

Reply:
I guess we can't access these flags if we place them in instance - as the receiver will be executed in executor (say, different JVM, or at least, serialized - deserialized even in test). So exposing them in object is really a hack only for testing.

Btw, `receivedAllRecords` is actually not needed to be exposed as outside, as we create new receiver instance and we can just set the default value to false. If we move it, we can either remove hack via adding more verbose code or keep the hack. 

I'll move out receivedAllRecords for now: please let me know if you prefer to remove the hack, then I'll make a change.

Reply:
> as the receiver will be executed in executor

ok, that makes sense; I was under the impression it was a driver thing since an instance of it was passed to the context, instead of one created somehow in the executor.

Anyway, what you have is fine.